src/toil/wdl/wdltoil.py

```diff
@@ -1578,6 +1587,46 @@ def map_over_typed_files_in_value(value: WDL.Value.Base, transform: Callable[[WD
         # All other kinds of value can be passed through unmodified.
         return value
 
+def map_over_files_in_value_check_null_type(value: WDL.Value.Base, original_value: WDL.Value.Base, expected_type: WDL.Type.Base) -> None:
+    """
+    Run through all nested values embedded in the given value and check that the null values are valid.
+
+    If a null value is found that does not have a valid corresponding expected_type, raise an error
+
+    (This is currently only used to check that coerced File types to null are valid, as in null elements have an equivalent File? type,
+    if this is to be used elsewhere, the error message should be changed to be implementation specific)
+
+    If one of the nested values is null but the equivalent nested expected_type is not optional, this is not allowed
+    :param value: WDL base value to check. This is the WDL value that has been transformed and has the null elements
+    :param original_value: The original WDL base value prior to the transformation. Only used for error messages
+    :param expected_type: The WDL type of the value
+    :return:
+    """
+    if isinstance(value, WDL.Value.File):
+        pass
+    elif isinstance(value, WDL.Value.Array) and isinstance(expected_type, WDL.Type.Array):
+        for elem, orig_elem in zip(value.value, original_value.value):
+            map_over_files_in_value_check_null_type(elem, orig_elem, expected_type.item_type)
+    elif isinstance(value, WDL.Value.Map) and isinstance(expected_type, WDL.Type.Map):
+        for pair, orig_pair in zip(value.value, original_value.value):
+            # The key of the map cannot be optional or else it is not serializable, so we only need to check the value
```

Comment:
Don't we need to check the key in case it is typed as a non-optional File but actually has a null value? Or is File not allowed as a key type?

Reply:
I don't think an optional file is allowed in a Map type because that has to be serializable to JSON. For example, MiniWDL complains:
```
(test.wdl Ln 6 Col 3) Map[File?,Int] may not be used in input/output a because the keys cannot be written to/from JSON
      Map[File?, Int] a = {"path/to/file": 1}
```